PeanutPlug.groovy

```diff
@@ -188,235 +145,311 @@ i/*		    input (
-				log "Power ${powerValue}"
-				state.powerValue = $powerValue
-				sendEvent(name: "power", value: powerValue)
-			}
-		} else {
+            
+            /*
+            http://www.zigbee.org/wp-content/uploads/2014/10/07-5123-06-zigbee-cluster-library-specification.pdf
+
+            Id     Name             Type     Range         Acc   Default M/O
+            0x0505 RMSVoltage       uint16 0x0000 – 0xFFFF R     0xFFFF O
+            0x0506 RMSVoltageMin    uint16 0x0000 – 0xFFFF R     0xFFFF O
+            0x0507 RMSVoltageMax    uint16 0x0000 – 0xFFFF R     0xFFFF O
+            0x0508 RMSCurrent       uint16 0x0000 – 0xFFFF R     0xFFFF O
+            0x0509 RMSCurrentMin    uint16 0x0000 – 0xFFFF R     0xFFFF O
+            0x050A RMSCurrentMax    uint16 0x0000 – 0xFFFF R     0xFFFF O
+            0x050B ActivePower      int16 -32768 – 32767   R     0x8000 O
+
+            Id     Name                 Type     Range           Acc   Default M/O
+            0x0600 ACVoltageMultiplier  uint16   0x0001 – 0xFFFF R     0x0001 O
+            0x0601 ACVoltageDivisor     uint16   0x0001 – 0xFFFF R     0x0001 O
+            0x0602 ACCurrentMultiplier  uint16   0x0001 – 0xFFFF R     0x0001 O
+            0x0603 ACCurrentDivisor     uint16   0x0001 – 0xFFFF R     0x0001 O
+            0x0604 ACPowerMultiplier    uint16   0x0001 – 0xFFFF R     0x0001 O
+            0x0605 ACPowerDivisor       uint16   0x0001 – 0xFFFF R     0x0001 O
+            */
+                                          
+            switch(descMap.attrInt)
+            {
+                case 0x0600:
+                    log.info "ACVoltageMultiplier $intVal"
+				    state.voltageMultiplier = intVal
+                    break
+                case 0x0601:
+				    log.info "ACVoltageDivisor $intVal"
+				    state.voltageDivisor = intVal
+                    break
+                case 0x0602:
+                	log.info "ACCurrentMultiplier $intVal"
+				    state.currentMultiplier = intVal
+                    break
+                case 0x0603:
+                    log.info "ACCurrentDivisor $intVal"
+				    state.currentDivisor = intVal
+                    break
+                case 0x0604:
+                    log.info "ACPowerMultiplier $intVal"
+				    state.powerMultiplier = intVal
+                    break
+                case 0x0605:
+                    log.info "ACPowerDivisor $intVal"
+				    state.powerDivisor = intVal
+                    break
+                case 0x0505:
+                    def voltageValue = intVal * getVoltageMultiplier()
+                    log "Raw Voltage: ${intVal}"
+                    log "Voltage Multiplier: ${getVoltageMultiplier()}"
+                    log "RMS Voltage: ${voltageValue}"
+				    state.voltage = voltageValue
+				    sendEvent(name: "voltage", value: voltageValue)
+                    break
+                case 0x0508:
+                    def currentValue = intVal * getCurrentMultiplier()

```

Comment:
I intentionally used a string format for current and voltage because a) it looks funny to display '0E-10' for 0 volts and b) it not only looks funny but it is misleading to display voltages to 15 decimal places.  These devices are no way near this accurate.
However, I am not married to this and am willing to discuss it.
Same goes for current, power, and energy.

Reply:
I'm not a groovy guru, so I looked at some decimal formatting functions.  It looks like round() does what you wanted (which I agree is useful) without changing the type of the numbers.  I wanted to keep the .state entries as pure numbers and not mix in string types.  Have a look at the change and see if you like it.

Reply:
Hi Tom,
Sorry to admit my ignorance, but I don't know how to look at the changes.  All I see is your original change without a round().
I would like to try the code before I agree.  Does this fix the '0E-10' problem?
Thanks,
Jim

Reply:
Hey Jim-

The change was in this commit (#2 of 3 as we discussed your feedback).  It isn't showing up on the github summary diff for some reason, but you can see the change here:  

https://github.com/jimcghee/Hubitat-PeanutPlug/pull/1/commits/9e554ff4d34dd532bce6af4336ae875462d3c549


Once we resolve all of the discussion, maybe it would make sense for me to just retract this pull request and submit a new one for you to sanity check.